pkg/controllermanager/controller/bastion/bastion_control.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// NewBastionReconciler creates a new instance of a reconciler which reconciles Bastions.
+func NewBastionReconciler(
+	logger logrus.FieldLogger,
+	gardenClient client.Client,
+	maxLifetime time.Duration,
+) reconcile.Reconciler {
+	return &reconciler{
+		logger:       logger,
+		gardenClient: gardenClient,
+		maxLifetime:  maxLifetime,
+	}
+}
+
+type reconciler struct {
+	logger       logrus.FieldLogger
+	gardenClient client.Client
+	maxLifetime  time.Duration
+}
+
+// Reconcile reacts to updates on Bastion resources and cleans up expired Bastions.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	logger := r.logger.WithField("bastion", request)
+
+	bastion := &operationsv1alpha1.Bastion{}
+	if err := r.gardenClient.Get(ctx, request.NamespacedName, bastion); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+
+		logger.Infof("Unable to retrieve object from store: %v", err)
+		return reconcile.Result{}, err
+	}
+
+	// do not reconcile anymore once the object is marked for deletion
+	if bastion.DeletionTimestamp != nil {
+		return reconcile.Result{}, nil
+	}
+
+	// fetch associated Shoot
+	shoot := gardencorev1beta1.Shoot{}
+	shootKey := kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name)
+	if err := r.gardenClient.Get(ctx, shootKey, &shoot); err != nil {
+		// This should never happen, as the shoot deletion is stopped unless all Bastions
+		// are removed. This is required because once a Shoot is gone, the Cluster resource
+		// is gone as well and without that, cleanly destroying a Bastion is not possible.
+		if apierrors.IsNotFound(err) {
+			logger.WithField("shoot", shoot.Name).Warn("Deleting bastion because target shoot is gone")
+			return reconcile.Result{}, client.IgnoreNotFound(r.gardenClient.Delete(ctx, bastion))
+		}
+
+		return reconcile.Result{}, fmt.Errorf("could not get shoot %v: %v", shootKey, err)
+	}
+
+	// delete the bastion if the shoot is marked for deletion
+	if shoot.DeletionTimestamp != nil {
+		logger.WithField("shoot", shoot.Name).Info("Deleting bastion because target shoot is in deletion")
+		return reconcile.Result{}, client.IgnoreNotFound(r.gardenClient.Delete(ctx, bastion))
+	}
+
+	// the Shoot for this bastion has been migrated to another Seed, we have to garbage-collect
+	// the old bastion (bastions are not migrated, users are required to create new bastions);
+	// equality is the correct check here, as the admission plugin already prevents Bastions
+	// from existing without a spec.SeedName being set. So it cannot happen that we accidentally
+	// delete a Bastion without seed (i.e. an unreconciled, new Bastion);
+	// under normal operations, shoots cannot be migrated to another seed while there are still
+	// bastions for it, so this check here is just a safety measure.
+	if !equality.Semantic.DeepEqual(shoot.Status.SeedName, bastion.Spec.SeedName) {
```

Comment:
This should be 
```suggestion
	if !equality.Semantic.DeepEqual(shoot.Spec.SeedName, bastion.Spec.SeedName) {
```

Otherwise, the `Bastion` will only be deleted when the shoot has already been migrated (which will prevent a graceful deletion of the bastion).